src/entities/nestedJoin/index.ts

```diff
@@ -0,0 +1,123 @@
+import { encodeFunctionData } from 'viem';
+import { Address, Hex } from '../../types';
+import { Token } from '../token';
+import { BALANCER_RELAYER, getPoolAddress } from '../../utils';
+import { Relayer } from '../relayer';
+import { parseNestedJoinCall } from './parseNestedJoinCall';
+import { TokenAmount } from '../tokenAmount';
+import { balancerRelayerAbi, bathcRelayerLibraryAbi } from '../../abi';
+import {
+    NestedJoinInput,
+    NestedJoinQueryResult,
+    NestedJoinCallInput,
+} from './types';
+import { getNestedJoinCalls } from './getNestedJoinCalls';
+import { doQueryNestedJoin } from './doQueryNestedJoin';
+import { NestedPoolState } from '../types';
+
+export class NestedJoin {
+    async query(
+        input: NestedJoinInput,
+        nestedPoolState: NestedPoolState,
+    ): Promise<NestedJoinQueryResult> {
+        const calls = getNestedJoinCalls(input, nestedPoolState);
+
+        const parsedCalls = calls.map((call) => parseNestedJoinCall(call));
+
+        const encodedCalls = parsedCalls.map((parsedCall) =>
+            encodeFunctionData({
+                abi: bathcRelayerLibraryAbi,
+                functionName: 'joinPool',
+                args: parsedCall.args,
+            }),
+        );
+
+        // append peek call to get bptOut
+        const peekCall = Relayer.encodePeekChainedReferenceValue(
+            Relayer.toChainedReference(
+                calls[calls.length - 1].outputReferenceKey,
+                false,
```

Comment:
Good question. I'm using non-temp refs based on the previous peek implementation, but I'm not sure it's needed. I'll check and let you know what I find out.

Reply:
Non-temp refs work as expected 👍 
Regarding reusing the existing ref, we could find a way to merge `encodeCalls` with `getPeekCalls`, but I believe it's better to keep both logics separated, since encodeCalls is reused on `query` and `buildCall`.